NIFI-14424 - Support for Confluent schema registry Protobuf wire form…

lkuchars opened the pull request:
Support for Confluent schema registry Protobuf wire format in ProtobufReader. UniversalProtobufReader class added.

<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary
This is the second PR for the [NIFI-14424](https://issues.apache.org/jira/browse/NIFI-14424) feature. It contains the new implementation of the Protobuf RecordReaderFactory.

[NIFI-14424](https://issues.apache.org/jira/browse/NIFI-14424)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ ] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

lkuchars committed:
NIFI-14424 Added StandardProtobufReader controller service.

markobean committed:
NIFI-14856 set the default status history timeframe in the UI graph to correespond to settings in nifi.properties

Signed-off-by: Pierre Villard <pierre.villard.fr@gmail.com>

This closes #10205.

dan-s1 committed:
NIFI-14848 Replaced switch statements with enhanced switch statements.

Signed-off-by: Pierre Villard <pierre.villard.fr@gmail.com>

This closes #10197.

lfrancke committed:
NIFI-14859: Fix PMD warnings by adding @Override annotations

Also updates the PULL_REQUEST_TEMPLATE.md to use the Maven Wrapper

Signed-off-by: Pierre Villard <pierre.villard.fr@gmail.com>

This closes #10202.

lkuchars commented on nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/ProtobufReader.java:
```diff
@@ -54,15 +54,17 @@
+    static final String EMPTY_PROTO = "google/protobuf/empty.proto";
+    static final String STRUCT_PROTO = "google/protobuf/struct.proto";
+    static final String TIMESTAMP_PROTO = "google/protobuf/timestamp.proto";
+    static final String WRAPPERS_PROTO = "google/protobuf/wrappers.proto";
```
Will change this back to private

pvillard31 committed:
NIFI-14860 Bump Spring to 6.2.10, Jetty to 12.0.25, Netty to 4.2.4.Final, and others (#10203)

- reactor-netty-http from 1.2.8 to 1.2.9 - https://github.com/reactor/reactor-netty/releases/tag/v1.2.9
- reactor-core and reactor-test from 3.7.8 to 3.7.9 - https://github.com/reactor/reactor-core/releases/tag/v3.7.9
- elasticsearch from 9.1.1 to 9.1.2 - https://github.com/elastic/elasticsearch-java
- janusgraph from 1.2.0-20250608-090535.869f248 to 1.2.0-20250812-182757.f621a1c - https://github.com/JanusGraph/janusgraph
- wire from 5.3.8 to 5.3.10 - https://github.com/square/wire/blob/master/CHANGELOG.md#version-5310
- parcefone from 3.0.0 to 3.1.0 - https://github.com/fluenda/ParCEFone/releases/tag/3.1.0
- oauth2-oidc-sdk from 11.27 to 11.27.1 - https://bitbucket.org/connect2id/oauth-2.0-sdk-with-openid-connect-extensions/src/master/CHANGELOG.txt
- AWS SDK v2 from 2.32.19 to 2.32.22 - https://github.com/aws/aws-sdk-java-v2/blob/master/CHANGELOG.md
- kotlin from 2.2.0 to 2.2.10 - https://github.com/JetBrains/kotlin/releases/tag/v2.2.10
- jetty from 12.0.24 to 12.0.25 - https://github.com/jetty/jetty.project/releases/tag/jetty-12.0.25
- netty from 4.2.3.Final to 4.2.4.Final - https://netty.io/news/2025/08/13/4-2-4.html
- spring from 6.2.9 to 6.2.10 - https://github.com/spring-projects/spring-framework/releases/tag/v6.2.10
- nimbus jose jwt from 10.4.1 to 10.4.2 - https://bitbucket.org/connect2id/nimbus-jose-jwt/src/master/CHANGELOG.txt

Signed-off-by: David Handermann <exceptionfactory@apache.org>

dariuszseweryn commented on nifi-extension-bundles/nifi-confluent-platform-bundle/nifi-confluent-platform-schema-api/src/main/java/org/apache/nifi/confluent/schema/ProtobufMessageSchema.java:
```diff
@@ -21,7 +21,7 @@
 /**
  * Basic abstraction over Protobuf schema message entity.
- * It contains only the fields that are needed for crude schema insights. It's useful when the
+ * Currently, it represents only the fields that are needed for crude schema insights. It's useful when the
```
"Currently" doesn't add much here. Adding hints about the future is not something relevant I think.

lkuchars replied:
Done

dariuszseweryn commented on nifi-extension-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/schema/access/AvroSchemaTextStrategy.java:
```diff
@@ -28,7 +28,6 @@
 import java.util.EnumSet;
 import java.util.Map;
 import java.util.Set;
-
```
Spurious edit I think? Revert?

lkuchars replied:
Done

dariuszseweryn commented on nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/ProtobufSchemaCompiler.java:
```diff
@@ -0,0 +1,290 @@
+    public ProtobufSchemaCompiler(final ComponentLog logger) {
+        this.logger = logger;
+        this.compiledSchemaCache = Caffeine.newBuilder()
+            .expireAfterWrite(1, TimeUnit.HOURS)
```
Why this is `expireAfterWrite`? AFAIU a given `SchemaIdentifier` will always point to the same `Schema` and as long as it is used we should not prune the cached value — therefore we should use `expireAfterRead`, no?

lkuchars replied:
Good catch. Thank you!!!

dariuszseweryn commented on nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/ProtobufSchemaCompiler.java:
```diff
@@ -0,0 +1,290 @@
+        // Validate that all schema identifiers end with .proto extension
+        validateSchemaDefinitionIdentifiers(schemaDefinition, true);
+
+        return executeWithTemporaryDirectory(tempDir -> {
```
We should create a temporary directory that is unique for the passed schema definition, otherwise parallel invocations can interact with each other. Ideally we could even have some locking for each SchemaDefinition, so parallel invocations for the same one will not interfere.

Edit: after more thorough reading — this comment is invalid

lkuchars replied:
Agreed to resolve

dariuszseweryn commented on nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/ProtobufSchemaCompiler.java:
```diff
@@ -0,0 +1,290 @@
+     * Performs recursive validation on all referenced schemas.
+     *
+     * @param schemaDefinition       the schema definition to validate
+     * @param isRootSchemaDefinition set to if schema definition is a root definition
```
```suggestion
     * @param isRootSchemaDefinition set to true if schema definition is a root definition, false otherwise
```

dariuszseweryn commented on nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/ProtobufSchemaCompiler.java:
```diff
@@ -0,0 +1,290 @@
+                    String.format("Schema identifier name '%s' must end with .proto extension", name));
+            }
+        } else {
+            throw new IllegalArgumentException("Schema identifier must have a name that ends with .proto extension");
```
A little too ambiguous for my taste
```suggestion
            throw new IllegalArgumentException("Schema identifier must have a name that ends with .proto extension, but it has none");
```

dariuszseweryn commented on nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/ProtobufSchemaCompiler.java:
```diff
@@ -0,0 +1,290 @@
+
+        // Write schema text to file
+        Files.write(schemaFile, schemaDefinition.getText().getBytes(), StandardOpenOption.CREATE, StandardOpenOption.WRITE);
+        logger.debug("Successfully wrote schema to file: {} (size: {} bytes)",
```
AFAIK there is a subtle difference between bytes length and `String.length()`
```suggestion
        logger.debug("Successfully wrote schema to file: {} (string length: {})",
```

dariuszseweryn commented on nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/ProtobufSchemaCompiler.java:
```diff
@@ -0,0 +1,290 @@
+     */
+    private String generateSchemaFileName(final SchemaDefinition schemaDefinition) {
+        String schemaFileName = schemaDefinition.getIdentifier().getName().orElseGet(
+            () -> String.valueOf(schemaDefinition.getIdentifier().getSchemaVersionId().orElse(1L))
```
Why not start at 0?
```suggestion
            () -> String.valueOf(schemaDefinition.getIdentifier().getSchemaVersionId().orElse(0L))
```

dariuszseweryn commented on nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/UniversalProtobufReader.java:
```diff
@@ -0,0 +1,304 @@
+    volatile PropertyValue schemaText;
+    volatile PropertyValue schemaName;
+    volatile PropertyValue schemaBranchName;
+    volatile PropertyValue schemaVersion;
```
Do these need to be `volatile`? Given they are set only in `OnEnabled`/`OnDisabled` they should be safe to access?

lkuchars replied:
That is true, but everywhere I see such variables defined in Controllers/Services they are mostly volatile defensively? I'll leave them like this.

dariuszseweryn commented on nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/test/java/org/apache/nifi/services/protobuf/TestProtobufRecordReader.java:
```diff
@@ -41,7 +41,6 @@
 import static org.junit.jupiter.api.Assertions.assertInstanceOf;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertNull;
-
```
revert?

dariuszseweryn commented on nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/test/java/org/apache/nifi/services/protobuf/converter/TestProtobufDataConverter.java:
```diff
@@ -79,6 +80,19 @@ public void testDataConverterForProto3() throws Descriptors.DescriptorValidation
         assertEquals(3, nestedRecord2.getValue("int32Option"));
     }
 
+    // Test to ensure that ProtobufDataConverter can handle nested messages.
```
this comment should be test name, no?

lkuchars replied:
fixed

pvillard31 committed:
NIFI-14837 Updated GitHub Registry Client to limit of Commits Retrieved (#10186)

Signed-off-by: David Handermann <exceptionfactory@apache.org>

ForThoseWhoComeAfter committed:
NIFI-14424 Added Protobuf Schema Support to Confluent Schema Registry (#10105)

- Added Schema Definition retrieval method to SchemaRegistry

Co-authored-by: Lukas Kucharski <lkuchars@gmail.com>
Co-authored-by: Michał Bobowski<221161517+ForThoseWhoComeAfter@users.noreply.github.com>
Signed-off-by: David Handermann <exceptionfactory@apache.org>

jrsteinebrey committed:
NIFI-14732 Deleted references to removed Encrypted Repositories (#10207)

Signed-off-by: David Handermann <exceptionfactory@apache.org>

exceptionfactory committed:
NIFI-14865 Fixed Config Verification for List Azure Processors

- Checked Listing Mode and created new Client Factory for configuration verification

Signed-off-by: Pierre Villard <pierre.villard.fr@gmail.com>

This closes #10209.

exceptionfactory committed:
NIFI-14866 Reduced use of external test files in Standard Processors (#10210)

- Removed unused certificate file
- Replaced JPEG with a short binary array
- Replaced XML files with inline string for SplitXml

Signed-off by: Otto Fowler <ottobackwards@gmai.com>

lkuchars commented:
Thank you for the review @dariuszseweryn

lkuchars commented:
Closing temporary PR
